### PR TITLE
Feature/reorganize classes

### PR DESCRIPTION
**Pull request by algo26-matthias**
Goal of this PR is to refactor the messy library into a much cleaner state. The big classes are broken down into smaller units, which perform one job only.
Thus the code should in general be:
- faster
- better readable
- better maintainable
- easier testable
- smaller in memory

Also the tests have been extended (more test cases, test more different things). Unit tests are still yet to come, though.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -13,7 +13,7 @@
     ],
     "require": {
         "ext-pcre": "*",
-        "php": ">=5.6.0"
+        "php": ">=7.2.0"
```
Comment (glensc) [inline composer.json]: "^7.2.0" rather for semver, php tries to follow that.

Comment (algo26-matthias) [reply composer.json]:
```suggestion
        "php": "^7.2.0"
```

Comment (algo26-matthias) [reply composer.json]: Thanks for the suggestion, I fixed it.

File: src/Punycode/FromPunycode.php [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        );
+    }
+
+
+    /**
+     * Checks, whether or not the provided string is a valid punycode string
+     * @param string $encoded
+     * @return boolean
+     */
+    private function validate($encoded)
```
Comment (algo26-matthias) [inline src/Punycode/FromPunycode.php]:
```suggestion
    private function validate($encoded): bool
```

File: src/ToIdn.php [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        $this->unicodeTransCoder = new TranscodeUnicode();
+        $this->punycodeEncoder = new ToPunycode($idnVersion);
+    }
+
+    /**
+     * @param string $host
+     *
+     * @return string
+     * @throws InvalidCharacterException
+     * @throws Exception\AlreadyPunycodeException
```
Comment (d00p) [inline src/ToIdn.php]: Nowhere else is the Exception subnamespace given for the @throws directive

Comment (glensc) [reply src/ToIdn.php]: I think as long as it's resolved correctly (`Exception` namespace imported), it  does not matter what style you use (importing classes or importing namespaces).

**Comment by d00p** [comment]
@glensc sure, it's just cosmetics, it's not prefixed with Exception\ anywhere else, that's why I was commenting it.

**Pull request merged**